File: src/agent_search_1688/tools/web/searxng.py
"""本地 SearXNG HTTP 客户端。"""

from __future__ import annotations

import json
from time import monotonic
from typing import Any
from urllib.parse import urlencode, urlparse
import urllib.error
import urllib.request


class SearXNGError(RuntimeError):
    pass


def validate_searxng_base_url(value: str) -> str:
    url = value.strip().rstrip("/")
    parsed = urlparse(url)
    if parsed.scheme not in {"http", "https"} or not parsed.netloc:
        raise ValueError("searxng_base_url 必须是完整的 HTTP(S) 地址")
    if parsed.query or parsed.fragment or parsed.path not in {"", "/"}:
        raise ValueError("searxng_base_url 不能包含路径、查询参数或片段")
    return url


def search_searxng(*, base_url: str, query: str, limit: int, timeout_seconds: int) -> dict[str, Any]:
    normalized_base_url = validate_searxng_base_url(base_url)
    normalized_query = query.strip()
    if not 2 <= len(normalized_query) <= 300:
        raise ValueError("query 长度必须在 2 到 300 个字符之间")
    if not 1 <= limit <= 20:
        raise ValueError("limit 必须在 1 到 20 之间")
    request = urllib.request.Request(f"{normalized_base_url}/search?" + urlencode({"q": normalized_query, "format": "json"}), headers={"Accept": "application/json", "User-Agent": "as1688/0.3.0"}, method="GET")
    started_at = monotonic()
    try:
        with urllib.request.urlopen(request, timeout=timeout_seconds) as response:
            raw = response.read(2_000_001)
    except urllib.error.HTTPError as exc:
        if exc.code in {401, 403}:
            raise SearXNGError("SearXNG 拒绝请求；请确认已启用 JSON 格式") from exc
        raise SearXNGError(f"SearXNG 请求失败（HTTP {exc.code}）") from exc
    except (urllib.error.URLError, TimeoutError, OSError) as exc:
        raise SearXNGError(f"无法连接本地 SearXNG：{exc}") from exc
    if len(raw) > 2_000_000:
        raise SearXNGError("SearXNG 响应过大")
    try:
        payload = json.loads(raw.decode("utf-8"))
    except (UnicodeDecodeError, json.JSONDecodeError) as exc:
        raise SearXNGError("SearXNG 没有返回有效 JSON；请启用 format=json") from exc
    return {"query": normalized_query, "results": parse_searxng_results(payload, limit), "meta": {"backend": "searxng", "duration_ms": round((monotonic() - started_at) * 1000)}}


def parse_searxng_results(payload: Any, limit: int) -> list[dict[str, str]]:
    raw_results = payload.get("results") if isinstance(payload, dict) else None
    if not isinstance(raw_results, list):
        raise SearXNGError("SearXNG JSON 缺少 results 列表")
    results: list[dict[str, str]] = []
    seen_urls: set[str] = set()
    for item in raw_results:
        if not isinstance(item, dict):
            continue
        url = item.get("url")
        if not isinstance(url, str) or not url.startswith(("http://", "https://")) or url in seen_urls:
            continue
        seen_urls.add(url)
        title, content, engine = item.get("title"), item.get("content"), item.get("engine")
        results.append({"title": title.strip() if isinstance(title, str) else "", "url": url, "snippet": content.strip() if isinstance(content, str) else "", "engine": engine.strip() if isinstance(engine, str) else ""})
        if len(results) >= limit:
            break
    return results

File: src/tools/web/searxng.ts
import { performance } from "node:perf_hooks";

export class SearxngError extends Error {
  override readonly name = "SearxngError";
}

export interface SearxngResult {
  readonly title: string;
  readonly url: string;
  readonly snippet: string;
  readonly engine: string;
}

export function validateSearxngBaseUrl(value: string): string {
  const normalized = value.trim().replace(/\/+$/u, "");
  let url: URL;
  try {
    url = new URL(normalized);
  } catch (error) {
    throw new Error("searxng_base_url 必须是完整的 HTTP(S) 地址", { cause: error });
  }
  if (!(["http:", "https:"] as string[]).includes(url.protocol) || !url.host) {
    throw new Error("searxng_base_url 必须是完整的 HTTP(S) 地址");
  }
  if (url.pathname !== "/" || url.search || url.hash) {
    throw new Error("searxng_base_url 不能包含路径、查询参数或片段");
  }
  return normalized;
}

export async function searchSearxng(options: {
  readonly baseUrl: string;
  readonly query: string;
  readonly limit: number;
  readonly timeoutSeconds: number;
  readonly fetchImpl?: typeof fetch;
}): Promise<Record<string, unknown>> {
  const baseUrl = validateSearxngBaseUrl(options.baseUrl);
  const query = options.query.trim();
  if (query.length < 2 || query.length > 300) throw new Error("query 长度必须在 2 到 300 个字符之间");
  if (!Number.isInteger(options.limit) || options.limit < 1 || options.limit > 20) throw new Error("limit 必须在 1 到 20 之间");
  const url = new URL(`${baseUrl}/search`);
  url.search = new URLSearchParams({ q: query, format: "json" }).toString();
  const started = performance.now();
  let response: Response;
  try {
    response = await (options.fetchImpl ?? fetch)(url, {
      headers: { Accept: "application/json", "User-Agent": "as1688/1.0.0" },
      signal: AbortSignal.timeout(options.timeoutSeconds * 1_000),
    });
  } catch (error) {
    throw new SearxngError(`无法连接本地 SearXNG：${safeError(error)}`, { cause: error });
  }
  if (!response.ok) {
    if (response.status === 401 || response.status === 403) {
      throw new SearxngError("SearXNG 拒绝请求；请确认已启用 JSON 格式");
    }
    throw new SearxngError(`SearXNG 请求失败（HTTP ${response.status}）`);
  }
  const raw = await readBoundedResponse(response, 2_000_000, "SearXNG 响应过大");
  let payload: unknown;
  try {
    payload = JSON.parse(new TextDecoder().decode(raw));
  } catch (error) {
    throw new SearxngError("SearXNG 没有返回有效 JSON；请启用 format=json", { cause: error });
  }
  return {
    query,
    results: parseSearxngResults(payload, options.limit),
    meta: { backend: "searxng", duration_ms: Math.round(performance.now() - started) },
  };
}

export function parseSearxngResults(payload: unknown, limit: number): SearxngResult[] {
  if (!isRecord(payload) || !Array.isArray(payload.results)) {
    throw new SearxngError("SearXNG JSON 缺少 results 列表");
  }
  const results: SearxngResult[] = [];
  const seen = new Set<string>();
  for (const item of payload.results) {
    if (!isRecord(item) || typeof item.url !== "string") continue;
    if (!(item.url.startsWith("http://") || item.url.startsWith("https://")) || seen.has(item.url)) continue;
    seen.add(item.url);
    results.push({
      title: typeof item.title === "string" ? item.title.trim() : "",
      url: item.url,
      snippet: typeof item.content === "string" ? item.content.trim() : "",
      engine: typeof item.engine === "string" ? item.engine.trim() : "",
    });
    if (results.length >= limit) break;
  }
  return results;
}

export async function readBoundedResponse(response: Response, maximum: number, message: string): Promise<Uint8Array> {
  if (!response.body) return new Uint8Array();
  const reader = response.body.getReader();
  const chunks: Uint8Array[] = [];
  let size = 0;
  while (true) {
    const { done, value } = await reader.read();
    if (done) break;
    size += value.byteLength;
    if (size > maximum) {
      await reader.cancel();
      throw new SearxngError(message);
    }
    chunks.push(value);
  }
  const merged = new Uint8Array(size);
  let offset = 0;
  for (const chunk of chunks) {
    merged.set(chunk, offset);
    offset += chunk.byteLength;
  }
  return merged;
}

function isRecord(value: unknown): value is Record<string, unknown> {
  return typeof value === "object" && value !== null && !Array.isArray(value);
}

function safeError(error: unknown): string {
  return error instanceof Error ? error.message.slice(0, 500) : String(error).slice(0, 500);
}
